HPCC-20950 Add ability to set or clear expire via fileservices

AttilaVamos opened the pull request:
Implement:
  - integer4 GetExpireDays(lfn)
  - SetExpireDays(lfn, expireDays)
  - ClearExpireDays(lfn)

functions in file services.

Extend spray_expire_test.ecl to test the 2 new functions.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [x] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
Tested locally.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-209.ca-central-1.compute.internal
Sha: ce578fa051c55ceef1eeea0c2d568c6083f5e68c
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |51 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |8 |8 |0 |24 sec  (00:00:24) |
|setup (thor) |8 |8 |0 |41 sec  (00:00:41) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |948 |948 |0 |310 sec  (00:05:10) |
|test (thor) |859 |859 |0 |774 sec  (00:12:54) |
|test (roxie) |1028 |1028 |0 |352 sec  (00:05:52) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  591 sec (00:09:51) |  105 sec (00:01:45) |   20 sec (00:00:20) |   16 sec (00:00:16) | 1940 sec (00:32:20) |   17 sec (00:00:17) | 2698 sec (00:44:58) |

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -3215,6 +3215,22 @@ class CDistributedFileBase : implements INTERFACE, public CInterface
+        return true;
+    }
+
+    virtual bool setExpire(int expireDays)
```
not sure point of boolean return value, since it's always true, why not void?

AttilaVamos replied:
Changed

jakesmith commented on ecllibrary/std/File.ecl:
```diff
@@ -1040,4 +1040,29 @@ EXPORT varstring GetDefaultDropZone() :=
+ * @param lfn           The name of the logical file.
+ 
+ 
+ * @return  	        Value of @expiredays properties of the file or -1 if that property not exists (never expire).
```
trivial: remove 2 blank lines between @param and @return, and comments could line up.

AttilaVamos replied:
Removed

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -3215,6 +3215,22 @@ class CDistributedFileBase : implements INTERFACE, public CInterface
+    virtual bool getExpire(int &expireDays)
+    {
+        expireDays = queryAttributes().getPropInt("@expireDays", -1);
+        return true;
```
I think it's okay that this function always succeeds (returns -1 if @expiryDays not set),
but no value in it always returning true, it would be better if method was:

```
virtual int getExpireDays()
```

AttilaVamos replied:
Changed

jakesmith commented on ecllibrary/std/File.ecl:
```diff
@@ -1040,4 +1040,29 @@ EXPORT varstring GetDefaultDropZone() :=
+ * Set the expire days property of the specified logical filename.
+ *
+ * @param lfn           The name of the logical file.
+ * @param expireDays    Number of days before file expired. Use -1 to remove 
```
see other comment re. ClearExpiryDays, and worth comment mentioning what 0 means.

AttilaVamos replied:
ClearExpireDays() function added, comment added to SetExpireDays() funtion parameter.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3008,3 +3010,31 @@ FILESERVICES_API void FILESERVICES_CALL fsGetDropZones(ICodeContext *ctx, size32
+    return expireDays;
+}
+
+FILESERVICES_API void FILESERVICES_CALL fsSetExpireDays(ICodeContext * ctx, const char *_lfn, int expireDays)
```
probably worth validating expiryDays is sensible, at least >=-1

AttilaVamos replied:
Validation added

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3008,3 +3010,31 @@ FILESERVICES_API void FILESERVICES_CALL fsGetDropZones(ICodeContext *ctx, size32
+    if (df)
+        df->setExpire(expireDays);
+    else
+        throw MakeStringException(0, "GetExpireDays: Could not find logical file %s", lfn.str());
```
trivial: MakeStringException should be deprecated really, makeStringExceptionV is preferable

AttilaVamos replied:
Changed

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -142,6 +142,8 @@ static const char * EclDefinition =
 "  varstring GetDefaultDropZone() : c,once,entrypoint='fsGetDefaultDropZone'; \n"
 "  dataset(FsDropZoneRecord) GetDropZones() : c,context,entrypoint='fsGetDropZones'; \n"
+"  integer4 GetExpireDays(const varstring lfn) : c,context,entrypoint='fsGetExpireDays'; \n"
+"  SetExpireDays(const varstring lfn, integer4 expireDays = -1) : c,context,entrypoint='fsSetExpireDays'; \n"
```
From the ECL users perspective, it would be clearer when wanting to remove expiry, if there was a ClearExpiryDays(const varstring lfn), and for SetExpiryDays to only accept a value of >= 0

AttilaVamos replied:
Done

jakesmith commented on dali/base/dadfs.hpp:
```diff
@@ -413,6 +413,8 @@ interface IDistributedFile: extends IInterface
     virtual bool isExternal() const = 0;
     virtual bool getSkewInfo(unsigned &maxSkew, unsigned &minSkew, unsigned &maxSkewPart, unsigned &minSkewPart, bool calculateIfMissing) = 0;
+    virtual bool getExpire(int &expireDays) = 0;
+    virtual bool setExpire(int expireDays) = 0;
```
see other comments re. return types

AttilaVamos replied:
Changed

jakesmith commented:
@AttilaVamos - looks good in general, a few mostly minor comments.

HPCCSmoketest commented:
Process of PR-14430, label: HPCC-20950-impr-master starts now.
The reason of this test is: Forced to re-test.
Commit ID: cce93d0ddc266ddd4790c8023c5e6840daac371b
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-129.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-129.ca-central-1.compute.internal
Sha: cce93d0ddc266ddd4790c8023c5e6840daac371b
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |8 |8 |0 |22 sec  (00:00:22) |
|setup (thor) |8 |8 |0 |40 sec  (00:00:40) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |948 |948 |0 |305 sec  (00:05:05) |
|test (thor) |859 |859 |0 |775 sec  (00:12:55) |
|test (roxie) |1028 |1028 |0 |351 sec  (00:05:51) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  600 sec (00:10:00) |  108 sec (00:01:48) |   20 sec (00:00:20) |   16 sec (00:00:16) | 1751 sec (00:29:11) |   18 sec (00:00:18) | 2521 sec (00:42:01) |

AttilaVamos commented:
@jakesmith updated, please re-review.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -142,6 +142,9 @@ static const char * EclDefinition =
 "  varstring GetDefaultDropZone() : c,once,entrypoint='fsGetDefaultDropZone'; \n"
 "  dataset(FsDropZoneRecord) GetDropZones() : c,context,entrypoint='fsGetDropZones'; \n"
+"  integer4 GetExpireDays(const varstring lfn) : c,context,entrypoint='fsGetExpireDays'; \n"
+"  SetExpireDays(const varstring lfn, integer4 expireDays = -1) : c,context,entrypoint='fsSetExpireDays'; \n"
```
Should remove default, which doesn't make sense now since SetExpireDays doesn't except -1

AttilaVamos replied:
Removed

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3008,3 +3011,46 @@ FILESERVICES_API void FILESERVICES_CALL fsGetDropZones(ICodeContext *ctx, size32
+    constructLogicalName(ctx, _lfn, lfn);
+    Linked<IUserDescriptor> udesc = ctx->queryUserDescriptor();
+    Owned<IDistributedFile> df = queryDistributedFileDirectory().lookup(lfn.str(),udesc, false,false, false, nullptr, defaultPrivilegedUser);
+    int expireDays = -1;
```
trivial: but now you're not filling a reference, no point in this line, may as well just return, i.e.:
```
   if (df)
       return df->getExpire();
   else
        throw makeStringExceptionV(0, "GetExpireDays: Could not find logical file %s", lfn.str());
}
```

AttilaVamos replied:
Changed

jakesmith commented on ecllibrary/std/File.ecl:
```diff
@@ -1040,4 +1040,38 @@ EXPORT varstring GetDefaultDropZone() :=
+ * Return the expire days property of the specified logical filename.
+ *
+ * @param lfn           The name of the logical file.
+ * @return  	        Value of @expiredays properties of the file or -1 if that property not exists (never expire)
```
trivial: comment start column not aligned.

jakesmith commented on ecllibrary/std/File.ecl:
```diff
@@ -1040,4 +1040,38 @@ EXPORT varstring GetDefaultDropZone() :=
+ * Set the expire days property of the specified logical filename.
+ *
+ * @param lfn           The name of the logical file.
+ * @param expireDays    Number of days before file expired. Use -1 to remove 
```
This comment needs updating (-1 no long valid here),
also worth adding comment to say what 0 means.

AttilaVamos replied:
Changed

jakesmith commented on ecllibrary/std/File.ecl:
```diff
@@ -1040,4 +1040,38 @@ EXPORT varstring GetDefaultDropZone() :=
+                        @expiredays property (never expire).
+ */
+
+EXPORT SetExpireDays(varstring lfn, integer4 expireDays = -1) := 
```
default value misleading/incorrect.

AttilaVamos replied:
Removed

jakesmith reviewed (changes requested):
@AttilaVamos - a couple of minor issues.

HPCCSmoketest commented:
Process of PR-14430, label: HPCC-20950-impr-master starts now.
The reason of this test is: Forced to re-test.
Commit ID: 36d48481fea750bbcd1894568f724ffc66bc74e4
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-213.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-213.ca-central-1.compute.internal
Sha: 36d48481fea750bbcd1894568f724ffc66bc74e4
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |21 sec  (00:00:21) |
|setup (thor) |9 |9 |0 |41 sec  (00:00:41) |
|setup (roxie) |19 |19 |0 |18 sec  (00:00:18) |
|test (hthor) |947 |947 |0 |309 sec  (00:05:09) |
|test (thor) |858 |858 |0 |775 sec  (00:12:55) |
|test (roxie) |1020 |1020 |0 |352 sec  (00:05:52) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  597 sec (00:09:57) |  106 sec (00:01:46) |   20 sec (00:00:20) |   17 sec (00:00:17) | 1752 sec (00:29:12) |   19 sec (00:00:19) | 2519 sec (00:41:59) |

AttilaVamos commented:
@jakesmith Updated, take a final check.

jakesmith commented:
@AttilaVamos - looks good, please squash

AttilaVamos committed:
HPCC-20950 Add ability to set or clear expire via fileservices

Implement:
  - integer4 GetExpireDays(lfn)
  - SetExpireDays(lfn, expireDays)
  - ClearExpireDays(lfn)

functions in file services.

Extend spray_expire_test.ecl to test the 3 new functions.

Tested locally.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Process of PR-14430, label: HPCC-20950-impr-master starts now.
The reason of this test is: Forced to re-test.
Commit ID: ecc6a46f291fb643165e6b6958e3b7a7c99128b8
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-59.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-59.ca-central-1.compute.internal
Sha: ecc6a46f291fb643165e6b6958e3b7a7c99128b8
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |6 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |21 sec  (00:00:21) |
|setup (thor) |9 |9 |0 |40 sec  (00:00:40) |
|setup (roxie) |19 |19 |0 |20 sec  (00:00:20) |
|test (hthor) |947 |947 |0 |300 sec  (00:05:00) |
|test (thor) |858 |858 |0 |774 sec  (00:12:54) |
|test (roxie) |1020 |1020 |0 |352 sec  (00:05:52) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  601 sec (00:10:01) |  105 sec (00:01:45) |   21 sec (00:00:21) |   16 sec (00:00:16) | 1744 sec (00:29:04) |   19 sec (00:00:19) | 2515 sec (00:41:55) |

jakesmith reviewed (approved):
@richardkchapman - ready to merge.

richardkchapman reviewed (approved):
Approved
